src/components/map/EventPopup.js

```diff
@@ -0,0 +1,216 @@
+            .filter(d => d.displayInReports)
+            .map(d => d.dataElement);
+
+        for (let d of displayElements) {
+            await loadOptionSet(d);
+        }
+
+        if (eventCoordinateField) {
+            const coordElement = programStageDataElements.find(
+                d => d.dataElement.id === eventCoordinateField
+            );
+
+            if (coordElement) {
+                eventCoordinateFieldName = coordElement.dataElement.name;
+            }
+        }
+    }
+
+    return { displayElements, eventCoordinateFieldName };
+};
+
+// Loads event data for the selected feature
+const loadEventData = async feature =>
+    feature ? apiFetch(`/events/${feature.id}`) : null;
+
+// Returns table rows for all display elements
+const getDataRows = (displayElements, dataValues, styleDataItem, value) => {
+    const dataRows = [];
+
+    // Include data element used for styling if not included below
+    if (
+        styleDataItem &&
+        !displayElements.find(d => d.id === styleDataItem.id)
+    ) {
+        dataRows.push(
+            <tr key={styleDataItem.id}>
+                <th>{styleDataItem.name}</th>
+                <td>{hasValue(value) ? value : i18n.t('Not set')}</td>
+            </tr>
+        );
+    }
+
+    // Include rows for each displayInReport data elements
+    displayElements.forEach(({ id, name, valueType, options }) => {
+        const { value } = dataValues.find(d => d.dataElement === id) || {};
+        let formattedValue = value;
+
+        if (valueType === 'COORDINATE' && value) {
+            formattedValue = formatCoordinate(value);
+        } else if (options) {
+            formattedValue = options[value];
+        } else if (!hasValue(value)) {
+            formattedValue = i18n.t('Not set');
+        }
+
+        dataRows.push(
+            <tr key={id}>
+                <th>{name}</th>
+                <td>{formattedValue}</td>
+            </tr>
+        );
+    });
+
+    if (dataRows.length) {
+        dataRows.push(<tr key="divider" style={{ height: 5 }} />);
+    }
+
+    return dataRows;
+};
+
+// Will display a popup for an event feature
+const EventPopup = props => {
+    const {
+        coordinates,
+        feature,
+        programStage,
+        eventCoordinateField,
+        styleDataItem,
+        onClose,
+    } = props;
+
+    const [eventData, setEventData] = useState();
+    const [dataElements, setDataElements] = useState({});
+    const { displayElements = [], eventCoordinateFieldName } = dataElements;
+
+    // Load data elements every time programStage is changed
+    useEffect(() => {
+        loadDataElements(programStage, eventCoordinateField).then(
+            setDataElements
+        );
+    }, [programStage]);
+
+    // Load event data every time a new feature is clicked
+    useEffect(() => {
+        loadEventData(feature).then(setEventData);
+    }, [feature]);
+
+    if (!coordinates || !feature || !eventData) {
```

Comment:
Should we move the `!feature || !eventData` check to be **inside** the Popup and show a loading spinner while the things are loading?  Otherwise there's a delay with nothing visible after the user clicks and before the popup displays

Reply:
Also, if coordinates is undefined nothing happens?  When will coordinates ever be undefined?

Reply:
I've done some refactoring here and made feature and coordinates required props of the popup. My original plan was to create a shared popup component as the event layer was created, but itwas not very clean. Now a new popup is created every time the user clicks on a new event. 

I tested with a spinner in the popup, but I didn't like the effect. It results in a fast blinking popup, as the popup with a spinner is much smaller. We don't know the size of the popup before the data is loaded.

If needed, I think we should solve a popup loading indicator in a separate issue.